tools/zh_segmenter/segment.py

```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/env python
+import argparse
+import pathlib
+import os
+import re
+import sys
+
+import tensorflow_text
+
+_DISALLOW_SEGMENTATION = re.compile(r'(?:\{[^\}]*\}|\\[a-z]|\s)+'.encode())
+
+_ZWSP = "\u200B"
+
+_MODEL_HANDLE = "https://tfhub.dev/google/zh_segmentation/1"
+
+default_input_path = pathlib.Path(__file__).resolve(
+).parent.parent.parent.joinpath("Translations/zh_CN.po")
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--input_path", default=default_input_path,
+                    help="Path to the input .po file")
+parser.add_argument("--output_path", help="Output path (default: in-place)")
+parser.add_argument("--separator", default=_ZWSP,
+                    help="Separator to use between segments")
+parser.add_argument(
+	"--debug_model", help="If this flag is given, simply segment the given string with | and prints it (without any processing)")
+args = parser.parse_args()
+
+_SEPARATOR = args.separator.encode()
+_SEGMENTER = tensorflow_text.HubModuleTokenizer(_MODEL_HANDLE)
+
+
+def _RecoverGaps(starts, ends):
+	"""Recovers gaps from the segmenter-produced start and end indices.
+
+	The segmenter may produce gaps around spaces, e.g. segment("hello world") => "hello|world"
+	"""
+	out_starts = []
+	out_ends = []
+	prev_end = 0
+	for start, end in zip(starts, ends):
+		if start != prev_end:
+			out_starts.append(prev_end)
+			out_ends.append(start)
+		out_starts.append(start)
+		out_ends.append(end)
+		prev_end = end
+	return out_starts, out_ends
+
+
+def _RunSegmenter(text):
+	_tokens, starts, ends = _SEGMENTER.tokenize_with_offsets(text)
+	starts, ends = _RecoverGaps(starts, ends)
+	return [text[i:j] for i, j in zip(starts, ends)]
+
+
+def _SegmentToParts(text):
+	result = []
+	cur = 0
+	for m in re.finditer(_DISALLOW_SEGMENTATION, text):
+		if cur != m.start():
+			result.extend(_RunSegmenter(text[cur:m.start()]))
+		result.append(text[m.start():m.end()])
+		cur = m.end(0)
+	if cur != len(text):
+		result.extend(_RunSegmenter(text[cur:]))
+	return result
+
+
+# We handle these at runtime, no need to mark it.
```

Comment:
I suppose alternatively we could simplify it so that this is also expected to be done at build time, thought that might not be idea for testing/Japanese?

Reply:
We could but let's keep it as is for now, can change later if we want to after thinking about it.